### ISSUE TITLE
Abusing JICS portals - causing instability for all campuses

### ISSUE THREAD

**Issue by Toasty66** [hyperschedule/hyperschedule#158]
If you're not aware, your tool is causing great stability problems with the Claremont SIS, as the course searches overwhelm the portals and subsequently the backend database.  For example, faculty were experiencing blank or truncated course lists, and the high load caused the process that synchronizes students into Sakai to deregister a good number of students from their courses.

This app looking very interesting, but in not discussing it with the individual campuses, it is effectively causing a denial of service.

**Comment by raxod502** [comment]
Hello @Toasty66. Thank you for your concern. I am the original author of Hyperschedule, although not the current maintainer (that would be @kwshi). Let me provide some context that may be helpful in finding a mutually satisfactory solution.

I originally developed Hyperschedule in Fall 2017, and it has been running continuously since then with no problems reported by any of the campuses. You are the first person, in fact, to ever report any trouble. Since no changes have been made in several years to how Hyperschedule retrieves data from the SIS, it seems likely to me that any problems are due to changes that have been made on the SIS side.

Hyperschedule imposes a fixed amount of load on the SIS, regardless of the number of students who use it. Requests are not made in parallel, and there is a minimum delay of at least 5 seconds between any two requests. Since performing a search through the SIS takes so long, there is typically only one search made about once every 30 to 60 seconds, or even less frequently.

I understand that the access pattern of Hyperschedule is different than typical student traffic, since Hyperschedule performs a wildcard search to retrieve all courses at the same time, rather than one at a time. This could impose additional system load, although I think the fact that students who would otherwise be using Portal directly will be accessing Hyperschedule's cached data should offset this cost to an extent.

The most important thing to know about Hyperschedule is that it is heavily relied upon by students at the Claremont Colleges, and has been for years. Monthly traffic exceeds 17,000 hits during course registration, according to analytics. I understand that Hyperschedule may not be a "supported service" from the point of view of the Claremont Colleges, but that doesn't matter from the point of view of the 1,000+ students who rely on it.

I think it would be wonderful if your department could devote some resources to working with the Hyperschedule development team in order to determine how SIS and Hyperschedule can continue to work together effectively to serve the students of the Claremont Colleges, as they have done successfully since 2017.

**Comment by Toasty66** [comment]
@raxod502 The first assumption is that this has been working fine since 2017, but that's not actually the case.  We've tracked many registration/course issues over the past few years that I'm now wondering if they were related to the excessive activity of the tool.  The college's vendor made a technology shift from cgi to java a while back on the portals, and that code base allows far more scaling and appears to be amplifying the impact from this tool.  I can't understate the impact this had yesterday, and the shear load this put on the entire system.  It impacted classes - that's probably not what was intended, but clearly is occurring. 

For example, three IP's from the same hosting company (QTS datacenters), combined for about 1/2 TB and 400K ingress packets per hour against just one of the college's portal.  As the investigation continues, similar activity has been found on the other portals.  The scraper sourcing is coming from other hosting too, such as AWS, suggesting that there are several iterations of the scraper in play here vs just one. It's one thing when it's just one scraper, and very different when it's multiples against every campus portal. 

It would behoove the developers of this tool to reach out to the HMC CIO and begin discussions. As you may or may not know, the current SIS is being replaced by Workday student, and work and enhancements against the existing system have been frozen.

**Comment by kwshi** [comment]
> For example, three IP's from the same hosting company (QTS datacenters), combined for about 1/2 TB and 400K ingress packets per hour against just one of the college's portal. As the investigation continues, similar activity has been found on the other portals. The scraper sourcing is coming from other hosting too, such as AWS, suggesting that there are several iterations of the scraper in play here vs just one. It's one thing when it's just one scraper, and very different when it's multiples against every 
campus portal.

This part is surprising to me, because as far as I can tell we _are_ just running one scraper.  We only run a single server process via Heroku, and our scraper only accesses courses through HMC's own Portal at portal.hmc.edu, not through any of the other campuses' portals--in fact, we simply have no code written at all that is capable of accessing any other campuses' portal.  So I'm not sure why you're observing "multiples against every campus portal".  Perhaps I'm misunderstanding how the SIS/portal infrastructure works, and accessing courses via the HMC Portal inadvertently propagates traffic to other colleges' portals too?  Also, could you please elaborate on why you're so confident that these impacts are coming from Hyperschedule, specifically?

- EDIT: I _may_ be mistaken about only using Heroku; I think @raxod502 is more familiar with how our scraper infrastructure is setup currently, so he'll have to correct me on that.

(It's possible that other instances of the scraper may be run locally by developers testing the [scraper script](https://github.com/MuddCreates/hyperschedule-api) locally, but those instances would not have any datacenter IPs and would _not_ be producing constant traffic.)

> It would behoove the developers of this tool to reach out to the HMC CIO and begin discussions. As you may or may not know, the current SIS is being replaced by Workday student, and work and enhancements against the existing system have been frozen.

Very well; I will contact the HMC registrar/CIS departments to see what can be done about this issue.

**Comment by raxod502** [comment]
> I may be mistaken about only using Heroku

No, you are not mistaken. @Toasty66, there is exactly one scraper deployed by the Hyperschedule team that I am aware of. It runs on Heroku with one instance, one process, one scraper. It accesses only the HMC portal, and makes only one type of request (performing a search through the web interface using Selenium, the same as a student would do).

If you're seeing more traffic, it's from a source of which I am entirely unaware. In fact, due to an internal bug in Hyperschedule, we are currently sending no traffic at all to SIS, and this has been the case for at least several hours—and I suspect it's really been much longer than that.

How do you know that the traffic is generated by Hyperschedule? My suspicion is that it isn't, since:
* we haven't sent any traffic at all lately, whereas you say you're seeing ongoing problems
* Hyperschedule has never changed the amount of traffic it's sent, whereas you say it was particularly bad yesterday
* Hyperschedule has never scraped any portal aside from HMC's, whereas you say you've seen traffic to multiple portals
* only one instance of Hyperschedule is deployed, whereas you say you've seen traffic from multiple IPs
* Hyperschedule has only ever been deployed on Heroku, whereas you say you've seen traffic from multiple different hosting providers

**Comment by Toasty66** [comment]
Don't know what to tell you. Perhaps, given this is on git, other enterprising students at TCC have setup additional scrapers, pointing at other portals, so now we have multiple instances continuously running. Here are the IPs that are the most prevalent. Let me know if any of these are your official ones. 

54.209.126.150 and 35.172.128.82 (both aws ec2)
216.154.201.228, 216.154.201.155, 216.154.201.157 (QTS datacenters)

**Comment by kwshi** [comment]
> Perhaps, given this is on git, other enterprising students at TCC have setup additional scrapers, pointing at other portals, so now we have multiple instances continuously running.

Perhaps, but even if that were the case, it wouldn't explain any of the traffic going to the other non-HMC portals--we simply _haven't written_ any code targeting those portals, so whoever is sending traffic to those portals must've written their own code.

Re. the IP addresses, I'm trying to look into that now to see what our IP addresses are; though I suspect that, since Heroku doesn't assign us a static IP, it's likely that our IP changes every time we relaunch/redeploy the process, and even if those IPs have been used before they are not the ones that are active now.  I'll see if I can make our scraper detect the IP of the host it's running on and print it to the logs, so that I can give you information on what our current scraper IP is and you can monitor the traffic associated with that.

**Comment by raxod502** [comment]
> Let me know if any of these are your official ones

You can see for yourself:

```
% dig hyperschedule.herokuapp.com

; <<>> DiG 9.10.6 <<>> hyperschedule.herokuapp.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 32617
;; flags: qr rd ra ad; QUERY: 1, ANSWER: 9, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;hyperschedule.herokuapp.com.	IN	A

;; ANSWER SECTION:
hyperschedule.herokuapp.com. 300 IN	CNAME	us-east-1-a.route.herokuapp.com.
us-east-1-a.route.herokuapp.com. 19 IN	A	3.233.183.24
us-east-1-a.route.herokuapp.com. 19 IN	A	34.199.176.68
us-east-1-a.route.herokuapp.com. 19 IN	A	34.234.209.139
us-east-1-a.route.herokuapp.com. 19 IN	A	34.193.9.51
us-east-1-a.route.herokuapp.com. 19 IN	A	52.200.54.195
us-east-1-a.route.herokuapp.com. 19 IN	A	52.70.167.158
us-east-1-a.route.herokuapp.com. 19 IN	A	52.200.77.100
us-east-1-a.route.herokuapp.com. 19 IN	A	52.20.96.224

;; Query time: 21 msec
;; SERVER: 2001:558:feed::1#53(2001:558:feed::1)
;; WHEN: Sat Oct 10 13:01:44 PDT 2020
;; MSG SIZE  rcvd: 216
```

**Comment by raxod502** [comment]
> Perhaps, given this is on git, other enterprising students at TCC have setup additional scrapers, pointing at other portals

Perhaps, but it's unclear to me why anybody would do this, given that all of the portals expose the same set of data (namely all the courses from all the schools).

**Comment by g-rocket** [comment]
Hi @Toasty66,

I'm Gavin, and I run another course-search webapp that (used to) scrape the HMC portal: https://github.com/hmc-portal2/hmc-scheduler (scheduler.yancey.io).

My code also scrapes the HMC Portal (or at least used to -- I think it broke a year or two ago and I haven't bothered fixing it).  And from the sounds of it, I'm much more agressive than Hyperschedule -- I do run requests in parallel, up to 30 (if I recall correctly) at a time.

Nonetheless, I just checked and 1/2 TB and 400k packets per hour is several orders of magnitude more than my program could ever generate, not to mention that mine would be sending traffic from a single residential IP and is only directed at the HMC portal.

I don't know what's sending all this traffic towards all the JCIS portals and I'm sorry you have to deal with it, but it's pretty clearly coming from neither Hyperschedule nor Scheduler.  Both of us only target the HMC portal, send much less / slower traffic, and only update occasionally.  To be honest, it sounds like someone might be trying to DOS the JCIS portal, but again that has nothing to do with this project.  I wish you the best of luck finding out what's actually going on.

**Comment by g-rocket** [comment]
@Toasty66, it looks like QTS [explicitly bans using their servers for a denial-of-service attack](https://www.qtsdatacenters.com/company/legal/acceptable-use-policy).  You might have better luck tracking down the culprit if you reported the issue to them directly; their contact info for reporting abuse is abuse@qtsdatacenters.com

**Comment by g-rocket** [comment]
@Toasty66 It looks like [AWS has a similar policy](https://aws.amazon.com/premiumsupport/knowledge-center/report-aws-abuse/) -- their contact looks to be ec2-abuse@amazon.com but the page I linked has some more details about how to report an incident.

**Comment by Toasty66** [comment]
Thanks all for the info.  It's clearly not something specifically designed to DoS the portals, as it's using the same/similar search logic as hyperschedule, including a prescriptive break between each search.  It very much feels like this code, or a close derivative.  Again, since it's available here, just about anyone could setup multiple scrapers and call it a day - if the intent is DoS, then there is nothing better than a resource-intensive search, amplified multiple times over.  It's one of the downsides to a publicly updated site scraper.

Abuse reports have already been filed for the hosts, and the college's are taking action to limit access to the course catalog from the public side of the portal and/or remove wildcard searching.  Should there be an opportunity, I'll capture the full search logic from these hosts and post it here for your information.

**Comment by raxod502** [comment]
> limit access to the course catalog from the public side of the portal and/or remove wildcard searching

If it's at all possible, please don't remove support for wildcard searches. This will break Hyperschedule, which will not be popular among the hundreds of students who rely on it to do their course registration.

If you must remove support, please provide an alternative way to extract data for use in Hyperschedule, and please keep the development team in the loop about what your plans are going forward.

**Comment by Toasty66** [comment]
@raxod502 We must ensure that the 5000+ undergrads can successfully register without disruption, which has not happened lately, including the registrars having to halt registration several times in August. I think you'd agree, that is not a great experience for the community, especially if caused by a scraper that benefits only a small percentage of the population. 

Did the developers of this tool ever seek out permission via the TCC SIS governance? Did you all work with the SIS teams to model the impact to the portals and backend systems? If development on the existing SIS system wasn't frozen, this sort of collaboration would be essential to determining the tools viability, including minimizing impact, and gaining the approvals for its use.  

If this tool is thought to be relevant for the new Workday SIS and/or the features should be incorporated in it, I'd recommend speaking soon with the SIS  implementation governance.

**Comment by raxod502** [comment]
> especially if caused by a scraper that benefits only a small percentage of the population

Okay, but—we've clearly established that Hyperschedule was not the cause of any of the disruption you're describing. In fact, your department hasn't even noticed the traffic from Hyperschedule, judging by the list of IPs that you posted.

> small percentage of the population

Hyperschedule is used almost universally by students at Harvey Mudd College, which is one of only five schools at the Claremont Colleges. I do not think Harvey Mudd constitutes a "small percentage of the population". As I mentioned in my first comments, we have analytics to back up this claim.

> Did the developers of this tool ever seek out permission via the TCC SIS governance?

I notified the office of the registrar at Harvey Mudd immediately, back in 2017, and they have known the details of Hyperschedule ever since. If Harvey Mudd did not follow the appropriate procedures in notifying your department, please take up your complaint with them.

> I'd recommend speaking soon with the SIS implementation governance

We'd be happy to work with the people working on the new system to find a mutually acceptable path forward. Could you please provide some contact information for this group?

Thanks for your patience.

**Comment by Toasty66** [comment]
@raxod502 The nature of AWS WC2 is that IP's change often, so we're running an experiment. As of 10:27am, the current offending AWS EC2 IP has been blocked on the HMC portal.  This IP had generated about 900 MB is data over the last two hours, and matches the wildcard searches we're seeing. 

Please let me know if your scraper has stopped working.  The plan is to run this test for the next hour, then determine next steps.

**Comment by kwshi** [comment]
@Toasty66:

- Our scraper has indeed stopped being able to connect to the HMC portal--so it seems likely that the address is actually associated with us.  (That our scraper is producing 900MB of data is still surprising to me, but if it is actually us, then I'm happy to pause our scraper for the time being until a solution is worked out.)

- Meanwhile, I just updated the app so that it reports its IP address (to its best knowledge) in its logs.  That means our app has just been redeployed, and the IP has likely changed from the one that has been blocked.  As of making this post, our scraper IP address should be 172.19.120.78.  This should help us confirm/identify exactly which of your requests are coming from us, and it would be great if you could further confirm whether that IP is actually generating all that traffic.

- I'm also working on adding some sort of identification for our requests, e.g., as part of the User-Agent string, so future requests can be identified without me needing to tell you an IP every time our app restarts.

Hope this helps.

**Comment by Toasty66** [comment]
@kwshi Given the AWS EC2 infrastructure, there may not be a match on IP. For one, your internal IP for the app will likely be a RFC1918 address e.g. 172.19.120.78, and then if there are load balancers, firewalls, or other items the public-facing IP may rotate. Somewhere in your dashboard will likely be an indication of its public facing IP (if you opted for a static one), otherwise it may be difficult to determine.  

What I can say is that at 11:40am, traffic picked up again from a new EC2 IP address of 18.212.15.211, same pattern, searching the course catalog. It's done about 200MB of data in just the last 60 minutes.  Here is an example of it's duty cycle against the HMC portal. 

I should mention, this is the exact pattern we saw against the other college portals, but coming from hosts outside of AWS EC2.  So either someone has deployed your code elsewhere, or they have taken inspiration from it. 

![HSCapture](https://user-images.githubusercontent.com/38083488/95786569-b5cd0880-0c8c-11eb-93f2-3b82b9347021.JPG)

**Comment by kwshi** [comment]
> @kwshi Given the AWS EC2 infrastructure, there may not be a match on IP. For one, your internal IP for the app will likely be a RFC1918 address e.g. 172.19.120.78, and then if there are load balancers, firewalls, or other items the public-facing IP may rotate. Somewhere in your dashboard will likely be an indication of its public facing IP (if you opted for a static one), otherwise it may be difficult to determine.

My mistake, didn't realize the IP I provided was internal.  I'm afraid we don't have a static IP, in particular, since we aren't even using AWS EC2 directly--only by way of Heroku, which dynamically launches and routes apps, etc., while providing no information about the IP address.  On this front, just for the sake of making future identification easier on your part, would it help for us to insert a string like "Hyperschedule scraper" into the User-Agent header?

> What I can say is that at 11:40am, traffic picked up again from a new EC2 IP address of 18.212.15.211, same pattern, searching the course catalog. It's done about 200MB of data in just the last 60 minutes. Here is an example of it's duty cycle against the HMC portal.

Curious.  I can't get to this at the moment, but I'll check in with the other maintainers and see if we can pause the scraper's Portal activity (or you're welcome to block the IP again) later today, and you can see how that affects the traffic in the next few days.  I'll update you when we do that.

> I should mention, this is the exact pattern we saw against the other college portals, but coming from hosts outside of AWS EC2. So either someone has deployed your code elsewhere, or they have taken inspiration from it.

That is strange, but I'm afraid I'm not sure how to help with that.  I've taken a look at [all the forks of the scraper project that GitHub knows about](https://github.com/MuddCreates/hyperschedule-api/network/members), and I think all of the forkers are Mudd students, which suggests to me that they would have no familiarity with any of the other campus portals.  And, as @raxod502 mentioned, since all the campus portals provide the exact same course data anyway, and since our API is public-facing and well-documented, whoever might be running these separate "inspired" instances on the other campus' portals have nothing to gain from doing so.  So, sorry I can't be of more help on this front.

**Comment by Toasty66** [comment]
Given how varied the IP addresses are, we would be engaging in wack-a-mole each time the app moves.  It would be far better to pause the scraper until you can have a discussion with the HMC CIO.

**Comment by raxod502** [comment]
While I'll leave the decision-making to @kwshi, I have to wonder: is it really the case that traffic from a single student is sufficient to bring down SIS?

Hyperschedule does not produce more traffic than a single student -- it makes search queries with no parallelism, using queries that are actually made in practice by students during registration.

If SIS truly cannot handle the load from Hyperschedule, which is basically equivalent to one student, then I have to wonder how it handles _more than one_ student making search queries during registration.

Something else must be going on, in my opinion.

**Comment by Toasty66** [comment]
@raxod502 Without the ability to look at what the backend is doing, what you are saying about a student vs the scraper is pure speculation. What I can say, from all the observed data, is it is different, and in is sustained in a way that isn't seen with students. Students don't exploit wildcard searching, which has a dramatic impact on system load.  The compounding of the scrapers (clearly not envisioned by the developers), adds even more because of how the system is shared between the undergrads. 

We need to first stop the exploit of the course search by all scrapers, harden the course search, then consider next steps.

**Comment by kwshi** [comment]
@Toasty66 Okay, I have upped our scraper's minimum timeout from the original 5 seconds to now 24 hours, a 17280x reduction.  Hopefully, that should all but eliminate the load of Hyperschedule's scraper on Portal.

For now, I decided against completely pausing the scraper, in the interest of keeping Hyperschedule data _somewhat_ up-to-date (just not receiving _live_ updates).

I've reached out to the HMC CIO and registrar to gather their thoughts on how to resolve this issue and how to move forward.  Let me know if that helps, as well as what next steps you have in mind to resolve this issue.

**Comment by g-rocket** [comment]
@Toasty66 I think it's *highly unlikely* the other scrapers have anything to do with this project, especially if they're running against other portals -- they're just using the most obvious method to scrape data from it.  For what it's worth, I created a scraper that works virtually identically to this one as part of my (entirely separate) course-scheduling web-app [a couple of months before this project existed](https://github.com/hmc-portal2/hmc-scheduler/commit/43cce596e515038ad1bc6fa5ec23d46ad19376eb).  It was running once-a-day from 2017 until I turned it off a few days ago in response to this issue (and because it doesn't really work anymore).

@kwshl, was Hyperschedule continuously scraping portal?  Why?  No wonder that generated a lot of load.  Once a day should be sufficient, except perhaps if you want to keep slots open up-to-date during registration, and even then once or twice an hour should be more than enough.

**Comment by g-rocket** [comment]
Historical side note: Several years back, HMC provided an API through lingk.io for apps like Hyperscheduler to retrieve this information without impacting the SIS database -- I believe it was updated with nightly data-dumps.  Eventually the API broke, and HMC switched from Lingk to Dell Boomi.  My understanding is HMC CIS was planning to remake the API, but pushback from TCC IT caused them to scrap that plan:

> After discussion with the cross-college IT governing body it was determined that we will 
not be able to support third-party applications attempting to retrieve bulk course 
registration data through the portal.  Please see the full response from the director of 
that group:
> 
> --
> 
> If real-time ... is a true requirement, I personally cannot see why we would pursue this.  
> It is likely to be difficult to make it perform well while not dragging down CX response 
> times.  (And it is precisely when CX is busy that the real-time enrollment data is most 
important.). I would not want to invest the effort in something that substantial when we 
expect WD Student in another year or so.  Realistically, even if we wanted to, this is 
not likely to be ready for FA 2020 registration, and we should have WD for FA 2021.  
So we’d be undertaking this effort for 1-2 terms.  Better to put effort into making WD 
Student a success.
>
> I am definitely partial towards providing real-time APIs.  I see the benefit, but in this 
case, I don’t see any compelling benefit/cost ratio.

It might be time to re-think that decision?  Admittedly the old API did have some issues with missing and incorrect data that forced apps like [hmc-scheduler](https://github.com/hmc-portal2/hmc-scheduler) to rely on *some* scraping to supplement the data, but having a working API would probably be the best solution for everyone moving forwards.

**Comment by Toasty66** [comment]
With the looming implementation of Workday student, nothing is being developed for the old SIS, be it alternative APIs, or incorporating new features.  That was a decision by the undergrads (including the council of presidents) so as to focus all investments on the new solution.

**Comment by kwshi** [comment]
> was Hyperschedule continuously scraping portal? Why? No wonder that generated a lot of load. Once a day should be sufficient, except perhaps if you want to keep slots open up-to-date during registration, and even then once or twice an hour should be more than enough.

Yes, it was originally scraping continuously (~a request a minute), for the reason you mentioned--because we want to keep up-to-date information on open slots.  I think I disagree that, during peak registration times, "once or twice an hour should be more than enough"; registration slots are 15 minutes each, and popular classes can fill up really quickly.  But I do agree that in the registration off-season, slowing the rate to ~twice a day should be sufficient.

> having a working API would probably be the best solution for everyone moving forwards.

> With the looming implementation of Workday student, nothing is being developed for the old SIS, be it alternative APIs, or incorporating new features.

I wholeheartedly agree with that sentiment.  But I think it doesn't preclude the idea of having an official API that works with the new Workday student.

**Comment by Toasty66** [comment]
@kwshi Workday is all about APIs vs the existing SIS that really has few/none.  I suspect then that there will be opportunity for this, but at the same time, it would be beneficial to participate in the UX to ensure that needed functionality is already there or on the development schedule. If it's built in, it benefits everyone.

**Comment by raxod502** [comment]
> Students don't exploit wildcard searching

This is not entirely correct. Wildcard search is a relatively known technique that has been used in practice by students during past course registrations. I know that I used it during my registration, after learning about it from another student. The advantage is that you don't have to wait a minute between each course, which is a valuable advantage since registration is so fast-paced.

**Comment by ssantichaivekin** [comment]
I just want to mention that Associated Students of Pomona College also has a scheduler [on their website](https://pomonastudents.org/courses/planner). The repository is available at https://github.com/aspc/mainsite. This scheduler is used by many Pomona students and is probably targeting Pomona's API (I cannot find on their repository which API the scheduler is targeting).

<img width="1440" alt="Screen Shot 2020-10-27 at 12 08 57 PM" src="https://user-images.githubusercontent.com/19219105/97259357-34a97000-184d-11eb-8b15-969316f482a5.png">

**Comment by ssantichaivekin** [comment]
The ASPC repository contains some code that seems to scrape the portal heavily. I have attached some code from their scraper below. I think we should ask them to join this conversation too.
```python
# In settings.py.example
# Set this value to be able to fetch data from Pomona's course API
COURSE_API_URL = ''
```
```python
# In import_courses.py
COURSES_URL = settings.COURSE_API_URL + 'courses/{0}/{1}'
...
class DepartmentEndpoint(object):
	def __init__(self, department_code, term_key):
		super(DepartmentEndpoint, self).__init__()
		self.url = COURSES_URL.format(term_key, department_code)

		try:
			self.courses_this_term = simplejson.load(urllib.urlopen(self.url))
		except simplejson.scanner.JSONDecodeError:
			self.courses_this_term = []
			logger.warning('could not access department URL ' + self.url)

class RequirementAreaEndpoint(object):
	def __init__(self, area_code, term_key):
		super(RequirementAreaEndpoint, self).__init__()
		self.url = COURSES_URL.format(term_key, area_code)

		try:
			self.courses_in_area = simplejson.load(urllib.urlopen(self.url))
		except simplejson.scanner.JSONDecodeError:
			self.courses_in_area = []
			logger.warning('could not access requirement area URL ' + self.url)
```
```python
# In refresh_enrollments.py
ENROLLMENTS_URL = settings.COURSE_API_URL + 'courses/%s'
...
    def handle(self, *args, **options):
        term = Term.objects.all().values_list('key', flat=True)[0]
        terms = get_all_terms(term)

        for t in terms:
            try:
                courses = simplejson.load(urllib.urlopen(ENROLLMENTS_URL % t))
```

**Comment by raxod502** [comment]
@Toasty66, we are currently in discussion with HMC's CIO, Joseph Vaughan. I am optimistic that we'll be able to work out a way for Hyperschedule to get the needed course data without having to go through the Portal web interface or impose very much database load, which will resolve the problems to the satisfaction of all parties involved.

**Comment by kwshi** [comment]
Updates since then:
- Portal course search now requires a student login.
- As such, we have stopped using the Portal scraper (as of ~2 months ago) & are now getting data via a ZIP file automatically sent to my school email address every 5 minutes.  If others come across this and want access to the data, reach out to me at kwshi@hmc.edu, and/or check out the API code now used to handle that data: https://github.com/MuddCreates/hyperschedule-api-go

Closing this issue since we no longer scrape Portal directly.
